### PR TITLE
Problem: missing json rpc of eth_feeHistory #685

### PR DESCRIPTION
**Pull request by leejw51crypto**
Solution: add feehistory api

### CODE REVIEW

**Comment by leejw51crypto** [comment]
will add unit test
and will limit count

because it's heavy task, can be used maliciously

**Comment by fedekunze** [comment]
@leejw51crypto let me know if you need support with this PR, I'd like to add it for the next release

**Comment by leejw51crypto** [comment]
ok

**Comment by leejw51crypto** [comment]
add unit test
rebased

**Comment by leejw51crypto** [comment]
fix golangci-lint

File: rpc/ethereum/backend/feebackend.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+		tendermintBlockResult, err := e.clientCtx.Client.BlockResults(e.ctx, &tendermintblock.Block.Height)
+		if err != nil {
+			e.logger.Debug("EthBlockFromTendermint block result not found", "height", tendermintblock.Block.Height, "error", err.Error())
+			return nil, err
+		}
+
+		onefeehistory := OneFeeHistory{}
+		processerr := e.processBlock(tendermintblock, &ethBlock, rewardPercentiles, tendermintBlockResult, &onefeehistory)
+		if processerr != nil {
+			return nil, processerr
```
Comment (fedekunze) [inline rpc/ethereum/backend/feebackend.go:178]:
```suggestion
			return nil, err
```

**Review by fedekunze — CHANGES_REQUESTED** [review]
thanks @leejw51crypto, please add more comments and clean up the logic based on the comments. Also if you could add a changelog entry

File: rpc/ethereum/backend/feebackend.go (hunk 1 above)
Comment (leejw51crypto) [reply rpc/ethereum/backend/feebackend.go:178]: ok, i'll modify

**Comment by leejw51crypto** [comment]
i'll modify

File: rpc/ethereum/backend/feebackend.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+	// set gasused ratio
+	gasLimitUint64 := (*ethBlock)["gasLimit"].(hexutil.Uint64)
+	gasUsedBig := (*ethBlock)["gasUsed"].(*hexutil.Big)
+	gasUsedString := gasUsedBig.ToInt().String()
+	gasusedfloat, _ := strconv.ParseFloat(gasUsedString, 64)
+	var gasUsedRatio float64
+	if gasLimitUint64 > 0 {
+		gasUsedRatio = gasusedfloat / float64(gasLimitUint64)
+	}
```
[the remark this answers is not in the document]
Comment (leejw51crypto) [reply rpc/ethereum/backend/feebackend.go:53]: yes, i'll return error if gaslimit is 0

File: rpc/ethereum/backend/feebackend.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+	thisGasUsedRatio := make([]float64, blockCount)
+
+	// fetch block
+	for blockID := blockStart; blockID < blockEnd; blockID++ {
+		index := int32(blockID - blockStart)
+		// eth block
+		ethBlock, err := e.GetBlockByNumber(rpctypes.BlockNumber(blockID), true)
+		if ethBlock == nil || err != nil {
+			return nil, err
+		}
```
[the remark this answers is not in the document]
Comment (leejw51crypto) [reply rpc/ethereum/backend/feebackend.go:160]: just err!=nil is enough

File: rpc/ethereum/backend/feebackend.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+		if err != nil {
+			return nil, err
+		}
+		blockEnd = int64(blockNumber)
+	}
+	userBlockCountInt := int64(userBlockCount)
+	const maxBlockCount = 100
+	if userBlockCountInt > maxBlockCount {
+		return nil, fmt.Errorf("FeeHistory user block count %d higher than %d", userBlockCountInt, maxBlockCount)
+	}
```
[the remark this answers is not in the document]
Comment (leejw51crypto) [reply rpc/ethereum/backend/feebackend.go:134]: ok

**Comment by leejw51crypto** [comment]
default history cap is 100 to prevent ddos attack,
can configured with app.toml
```
# FeeHistoryCap sets the global cap for total number of blocks that can be fetched
feehistory-cap = 100
```

File: rpc/ethereum/backend/backend.go [hunk 5]
```diff
@@ -889,6 +892,11 @@ func (e *EVMBackend) RPCFilterCap() int32 {
 	return e.cfg.JSONRPC.FilterCap
 }
 
+// RPCFeeHistoryCap is the limit for total number of blocks that can be fetched
+func (e *EVMBackend) RPCFeeHistgoryCap() int32 {
```
Comment (fedekunze) [inline rpc/ethereum/backend/backend.go:896]:
```suggestion
func (e *EVMBackend) RPCFeeHistoryCap() int32 {
```
```suggestion
func (e *EVMBackend) RPCFeeHistgoryCap() int32 {
```

File: rpc/ethereum/backend/feebackend.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+	// set gasused ratio
+	gasLimitUint64 := (*ethBlock)["gasLimit"].(hexutil.Uint64)
+	gasUsedBig := (*ethBlock)["gasUsed"].(*hexutil.Big)
+	gasusedfloat, _ := new(big.Float).SetInt(gasUsedBig.ToInt()).Float64()
+	var gasUsedRatio float64
+	if gasLimitUint64 > 0 {
+		gasUsedRatio = gasusedfloat / float64(gasLimitUint64)
+	} else {
+		return fmt.Errorf("gasLimit of block height %d should be bigger than 0 , current gaslimit %d", blockHeight, gasLimitUint64)
+	}
```
Comment (fedekunze) [inline rpc/ethereum/backend/feebackend.go:57]:
```suggestion

	if gasLimitUint64 <= 0 {
		return fmt.Errorf("gasLimit of block height %d should be bigger than 0 , current gaslimit %d", blockHeight, gasLimitUint64)
	}
	
	gasUsedRatio := gasusedfloat / float64(gasLimitUint64)
```
```suggestion
	var gasUsedRatio float64
	if gasLimitUint64 > 0 {
		gasUsedRatio = gasusedfloat / float64(gasLimitUint64)
	} else {
		return fmt.Errorf("gasLimit of block height %d should be bigger than 0 , current gaslimit %d", blockHeight, gasLimitUint64)
	}
```

File: rpc/ethereum/backend/feebackend.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+		txGasUsed := uint64(eachTendermintTxResult.GasUsed)
+		for _, msg := range tx.GetMsgs() {
+			ethMsg, ok := msg.(*evmtypes.MsgEthereumTx)
+			if !ok {
+				continue
+			}
+			tx := ethMsg.AsTransaction()
+			reward := tx.EffectiveGasTipValue(blockBaseFee)
+			sorter[i] = txGasAndReward{gasUsed: txGasUsed, reward: reward}
+			break
```
Comment (fedekunze) [inline rpc/ethereum/backend/feebackend.go:91]: I don't understand the purpose of this break here, are we only supporting 1 `MsgEthereumTx` message?

File: rpc/ethereum/backend/feebackend.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+		}
+		targetOneFeeHistory.Reward[i] = chosenReward
+	}
+
+	return nil
+}
+
+func (e *EVMBackend) FeeHistory(userBlockCount rpc.DecimalOrHex, /* number blocks to fetch, maximum is 100 */
+	lastBlock rpc.BlockNumber, /* the block to start search , to oldest */
+	rewardPercentiles []float64) /* percentiles to fetch reward */ (*rpctypes.FeeHistoryResult, error) {
```
Comment (fedekunze) [inline rpc/ethereum/backend/feebackend.go:120]:
```suggestion
func (e *EVMBackend) FeeHistory(
	userBlockCount rpc.DecimalOrHex, // number blocks to fetch, maximum is 100
	lastBlock rpc.BlockNumber, // the block to start search , to oldest
	rewardPercentiles []float64, // percentiles to fetch reward
)  (*rpctypes.FeeHistoryResult, error) {
```

File: rpc/ethereum/backend/feebackend.go [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+			return nil, err
+		}
+
+		// copy
+		thisBaseFee[index] = (*hexutil.Big)(onefeehistory.BaseFee)
+		thisGasUsedRatio[index] = onefeehistory.GasUsedRatio
+		for j := 0; j < rewardcount; j++ {
+			reward[index][j] = (*hexutil.Big)(onefeehistory.Reward[j])
+		}
+
```
Comment (fedekunze) [inline rpc/ethereum/backend/feebackend.go:187]:
```suggestion
```

File: rpc/ethereum/types/types.go [hunk 10]
```diff
@@ -79,3 +81,9 @@ type SignTransactionResult struct {
 	Raw hexutil.Bytes         `json:"raw"`
 	Tx  *ethtypes.Transaction `json:"tx"`
 }
+
+type OneFeeHistory struct {
+	BaseFee      *big.Int   /* base fee  for each block */
+	Reward       []*big.Int /* each element of the array will have the tip provided to miners for the percentile given */
+	GasUsedRatio float64    /* the ratio of gas used to gas limit for each block */
```
Comment (fedekunze) [inline rpc/ethereum/types/types.go:88]:
```suggestion
	BaseFee      *big.Int   // base fee  for each block
	Reward       []*big.Int // each element of the array will have the tip provided to miners for the percentile given
	GasUsedRatio float64    // the ratio of gas used to the gas limit for each block
```

**Review by fedekunze — COMMENTED** [review]
LGTM minor comments

**Review by fedekunze — APPROVED** [review]

**Pull request merged**